### PR TITLE
MRG: FIX time offset bug in raw.crop for Annotations

### LINKED ISSUE #4697: Misalignment of annotations after cropping raw
Hi,

I have an EEG dataset of one hour length covering several experiments. To mark the pauses between particular experiments as bad, I detect event codes marking their start and end point. After getting the corresponding time, I add it to the annotations variable and save it for further processing. The code is quite the same as in the [example](https://martinos.org/mne/dev/auto_tutorials/plot_artifacts_correction_rejection.html?highlight=blink%20annotation#marking-bad-raw-segments-with-annotations).

```
    raw = mne.io.read_raw_brainvision(raw_fname, montage=montage, preload=False)
    events = mne.find_events(raw, stim_channel='STI 014')

#%% Marking pauses as bad
    onset = np.array([events[idx,0] for idx in range(len(events[:, 2])) if events[idx, 2] > 80 and events[idx, 2] < 90])+1
    onset = np.insert(onset, 0, 0)
    end = np.array([events[idx,0] for idx in range(len(events[:, 2])) if events[idx, 2] > 70 and events[idx, 2] < 80])-1
    if events[0,2] < 70:
        end = np.concatenate([[events[0,0]-1], end, [len(raw)]])
    else:
        end = np.concatenate([end, [len(raw)]])

    duration = (end-onset)/raw.info['sfreq']
    onset = onset/raw.info['sfreq']
    description = ['bad']*len(onset)

    raw.annotations = mne.Annotations(onset, duration, description,
                                      orig_time=raw.info['meas_date'])

    raw.save(raw_fname+'.raw.fif.gz')
```

In a subsequent step, I load the EEG data, crop it and concatenate it. That results in a continous stream containing only the occurences of a certain experiment. 

```
    raw = mne.io.read_raw_fif(filename, preload=False)
    events = mne.find_events(raw, stim_channel='STI 014')

#%% cropping - assessing the beginning of particular experiments
    event_id = experiments['resting']['event_id']
    color = experiments['resting']['color']

# indices of the start and end triggercodes are detected, +1 and -1 to take a neighboring sample
    times_start = events[:, 0][np.where(events[:, 2] == event_id['start'])]-1
    times_end = events[:, 0][np.where(events[:, 2] == event_id['end'])]+1

    old_raw = raw.copy()
    new_raw = raw.copy().crop(raw.times[times_start[0]],
                              raw.times[times_end[0]])
    for idx in range(1, len(times_start)):
        mne.concatenate_raws([new_raw,
                              raw.copy().crop(raw.times[times_start[idx]],
                                              raw.times[times_end[idx]])])
# data is loaded, now the samples are loaded to memory
    raw = new_raw
    new_raw.close()
    raw.load_data()
    events = mne.find_events(raw, stim_channel='STI 014')
```

That is probably not the most elegant way of doing it.

The main problem is that after cropping the annotations are not where they are supposed to be. The following pictures show what I mean. 

issue_annotations_oldraw

issue_annotations_newraw

The annotations appear now at a totally different position which does not even fit to the onset values in the annotations variable:

```
old_raw.annotations.onset
Out[52]: 
array([    0.        ,   165.36399841,   609.80700684,   684.23101807,
        1065.14794922,  1497.01794434,  1589.45605469,  1963.6159668 ,
        2370.41699219,  2510.62597656,  2877.16308594,  3300.84399414,
        3431.56201172,  3873.2409668 ,  3942.5090332 ])

old_raw.first_samp
Out[53]: 0

new_raw.annotations.onset
Out[54]: array([ 437.35700684,  511.78101807])

new_raw.first_samp
Out[55]: 172450
```

I am not sure were the mistake happens. Probably, I misunderstood how the `orig_time` parameter works.
What is your suggestion?

### PR DESCRIPTION
**Pull request by agramfort**
closes #4697

### CODE REVIEW

File: mne/io/base.py [hunk 1]
```diff
@@ -1596,7 +1596,6 @@ def crop(self, tmin=0.0, tmax=None):
 
         if self.annotations is not None:
             annotations = self.annotations
-            annotations.onset -= tmin
```
Comment (agramfort) [inline mne/io/base.py]: since annotations times are all in seconds wrt to absolute time this is no need subtract anything when cropping.

**Comment by agramfort** [comment]
@jaeilepp if you have a minute to look it's great

@larsoner don't resist to look also ... :)

**Comment by larsoner** [comment]
AppVeyor is unhappy, failure looks legitimate

https://ci.appveyor.com/project/mne-tools/mne-python/build/1.0.1141#L484

File: mne/tests/test_annotations.py [hunk 2]
```diff
@@ -86,8 +87,22 @@ def test_annotations():
...
+    events = mne.find_events(raw)
+    onset = events[events[:, 2] == 1, 0] / raw.info['sfreq']
+    duration = np.ones_like(onset)  # 1s
+    description = ['bad %d' % k for k in range(len(onset))]
+    raw.annotations = mne.Annotations(onset, duration, description,
+                                      orig_time=raw.info['meas_date'])
+
+    raw_cropped = raw.copy().crop(raw.times[-1] / 2. + 2., None)
+    assert_array_equal(raw_cropped.annotations.description,
+                       raw.annotations.description[len(onset) // 2:])
```
Comment (larsoner) [inline mne/tests/test_annotations.py]: shouldn't we also check that the `onset` is still the same (`assert_allclose` at least)?

How about checking that if you add annotations, crop, and concatenate, that the annotations are equal?

**Comment by stfnrpplngr** [comment]
I get an error running the new test:

```
  File "<ipython-input-11-4d29b00c1f82>", line 2, in <module>
    raw.annotations.description[len(onset) // 2:])

  File "C:\Anaconda3\lib\site-packages\numpy\testing\utils.py", line 813, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')

  File "C:\Anaconda3\lib\site-packages\numpy\testing\utils.py", line 694, in assert_array_compare
    raise AssertionError(msg)

AssertionError: 
Arrays are not equal

(shapes (0,), (3,) mismatch)
 x: array([], 
      dtype='<U5')
 y: array(['bad 3', 'bad 4', 'bad 5'], 
      dtype='<U5')
```

Obviously, the annotations are not handed over from `raw` to `raw_cropped`.

**Comment by larsoner** [comment]
And AppVeyor has a [different error](https://ci.appveyor.com/project/mne-tools/mne-python/build/1.0.1141#L447). It's odd that Travis is happy, and tests pass locally for me, though. Maybe some floating point bug?

**Comment by agramfort** [comment]
@stfnrpplngr you get this error just by running the unit test? on what system?

**Comment by larsoner** [comment]
I agree it's probably a windows thing. FWIW I used UTC timecodes here:

https://github.com/mne-tools/mne-python/blob/master/mne/io/meas_info.py#L65

It seemed to fix some inconsistencies in handling.

**Comment by agramfort** [comment]
ok thx. I may have time to look at this next week

**Comment by larsoner** [comment]
Tests pass for me on a Windows 10 VM. A bit later I can try on a Windows 7 VM

**Comment by larsoner** [comment]
Tests pass on my Windows 7 VM (Python 3.6). @stfnrpplngr what is your `mne.sys_info()`?

**Comment by stfnrpplngr** [comment]
```
In [29]:  mne.sys_info()
Platform:      Windows-7-6.1.7601-SP1
Python:        3.6.2 |Anaconda custom (64-bit)| (default, Jul 20 2017, 12:30:02) [MSC v.1900 64 bit (AMD64)]
Executable:    C:\Anaconda3\pythonw.exe
CPU:           Intel64 Family 6 Model 42 Stepping 7, GenuineIntel: 8 cores
Memory:        16.0 GB

mne:           0.16.dev0
numpy:         1.11.2 {blas=mkl_core_dll, lapack=mkl_core_dll}
scipy:         0.18.1
matplotlib:    2.0.2

sklearn:       0.18.1
nibabel:       Not found
mayavi:        Not found
pycuda:        Not found
skcuda:        Not found
pandas:        0.20.1
```

File: mne/tests/test_annotations.py (hunk 2 above)
Comment (larsoner) [reply mne/tests/test_annotations.py]: FWIW the test suggested here fails on this PR.

**Comment by larsoner** [comment]
Here is your failing test:

https://travis-ci.org/mne-tools/mne-python/jobs/326418269#L2674

**Comment by larsoner** [comment]
Okay @agramfort I fixed the code I think. Ready for review/merge from my end.

@stfnrpplngr can you try running the updated tests, and also see if this fixes your problem from #4697?

**Comment by agramfort** [comment]
I don't think this is working as expected:

```
import numpy as np
import mne
from mne.datasets import sample

data_path = sample.data_path()
raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
raw = mne.io.read_raw_fif(raw_fname, preload=True)

onset = np.array([ 0.47058824,  2.49773765,  6.67873287,  9.15837097])
description = ['BAD'] * 4
duration = np.array([ 0.89592767,  2.13574672,  2.09954739,  0.48868752])
annotations = mne.Annotations(onset=onset, description=description,
                              duration=duration)
raw.annotations = annotations

raw.copy().crop(0, 20).plot(show_first_samp=True)

r = mne.concatenate_raws([raw.copy().crop(0, 4), raw.copy().crop(4, 8), raw.copy().crop(8, 20)])
r.plot(show_first_samp=True)
```

in the end you should see pretty much the same plots with also the annots for edge artifacts but that's it. Or I am missing something...

**Comment by larsoner** [comment]
Should be fixed

**Comment by agramfort** [comment]
works like a charm. Updated what's new.

**Pull request merged**